hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,216 @@
+# HAL overhaul : SPI
+
+## Description
+
+This updates targets the SPI HAL as the first step of a wide plan for all HAL APIs.
+
+## Motivation
+
+This work is part of the effort to bring well defined specification to APIs in `mbed-os`. Some users have shown interest in some addition to the supported features of this API ([bit ordering selection #4789](https://github.com/ARMmbed/mbed-os/issues/4789)).  
+Analysis of previous work also revealed some inconsistencies in the current SPI HAL API that this RFC aims at fixing. Those inconsistencies are listed in the next chapter.  
+Finally, `mbed-os` has grown a lot its supported target count and some of the oldest one could benefit quite significantly in maintenance cost from a factorisation of their code base.
+
+### Inconsistencies
+
+- The type used in block transfer functions vary between `(const) void*` and `(const) char*`.
+- Some method have a parameter to set the width of the symbols while it is part of the `spi_format` arguments.
+- Some method provide a `fill_symbol` parameter as a `char` which would not allow to set a symbol whose width is over 8bits.
+- Some method that allow to read more symbol than they write miss an argument to set the `fill_symbol`.
+- `spi_master_transfer` takes a `uint32_t` instead of a function pointer type for its `handler` argument.
+- Some method have misleading naming (e.g. `spi_slave_receive` and `spi_slave_read`).
+- The slave and master API are "almost" identical. There is no `spi_mater_read` nor `spi_master_receive`.
+- `spi_format` should use an enum for the mode.
+
```

Comment:
We know about some devices behaving in such way. For example the [PN512](https://www.nxp.com/docs/en/data-sheet/PN512.pdf#G1001371415) (an NFC frontend) is one of them.

Reply:
From the PN512 spec you pointed to, I don' t understand how you will use different Tx and Rx length. Can you explain ?

Reply:
The example of the NFC PN512 front end is just a case where you would have 1 byte of padding which means you will use length of N+1 to get N bytes, with a byte padding on the last Tx and a useless byte on the first Rx. But driver doesn't need a different tx_len and rx_len. 
If I look at PN512 driver here  https://github.com/wsrf2009/contactless-driver/blob/master/mod/src/pn512.c it actually uses a spi_write_then_read API.

Reply:
Actually, having the flexibility of only reading/writing the bytes you need (by skipping some leading/trailing bytes) can help prevent having to carry an extra copy or split a transfer into two or there separate ones.

For instance if I have buffers A and B and want to carry out this transfer:
```
TX: A0 A1 X
RX: X  B0 B1
```

If TX/RX lengths are the same and I can't skip bytes, I'll have to either:
1 - Copy my buffer A into a three byte long buffer, do the transfer and store results in the same buffer, then copy two bytes to my buffer B
2 - Split my transfer into three transfers

When buffers start being quite big this can be a bit problematic in terms of memory/processing costs. Splitting the transfer has a performance impact.

Reply:
@donatieng With the proposed API you would still have to allocated a 3 bytes buffer for reception as there is no "skip" parameter. Is that Ok or should we consider adding this new arguments ?
(( we might also consider packing arguments in a new structures as it would start to be a bit crowded in this arglist )).

Reply:
In my experience using the current API, I've found that either it's a small transfer, in which case a couple of extra bytes aren't significant, or it's a big transfer, in which case splitting into multiple parts isn't significant.